src/kokkos/KokkosCore/HistoContainer.h

```diff
@@ -0,0 +1,339 @@
+  //   Kokkos::parallel_for("histo_add",Kokkos::RangePolicy<KokkosExecSpace>(0,totbins()),
+  //     KOKKOS_LAMBDA(const int& i){
+  //       Kokkos::atomic_fetch_add(histo(0).off + i,co(0).off[i]);
+  //     });
+  // }
+
+  static KOKKOS_INLINE_FUNCTION uint32_t atomicIncrement(Counter &x) {
+    return Kokkos::atomic_fetch_add(&x, 1);
+  }
+
+  static KOKKOS_INLINE_FUNCTION uint32_t atomicDecrement(Counter &x) {
+    return Kokkos::atomic_fetch_sub(&x, 1);
+  }
+
+  KOKKOS_INLINE_FUNCTION void countDirect(T b) {
+    assert(b < nbins());
+    atomicIncrement(off[b]);
+  }
+
+  KOKKOS_INLINE_FUNCTION void fillDirect(T b, index_type j) {
+    assert(b < nbins());
+    auto w = atomicDecrement(off[b]);
+    assert(w > 0);
+    bins[w - 1] = j;
+  }
+
+  KOKKOS_INLINE_FUNCTION int32_t bulkFill(Kokkos::View<AtomicPairCounter*,KokkosExecSpace> apc, index_type const *v, uint32_t n) {
+    auto c = apc(0).add(n);
+    if (c.m >= nbins())
+      return -int32_t(c.m);
+    off[c.m] = c.n;
+    for (uint32_t j = 0; j < n; ++j)
+      bins[c.n + j] = v[j];
+    return c.m;
+  }
+
+  KOKKOS_INLINE_FUNCTION void bulkFinalize(Kokkos::View<AtomicPairCounter*,KokkosExecSpace> const apc) {
+    off[apc(0).get().m] = apc(0).get().n;
+  }
+
+  KOKKOS_INLINE_FUNCTION void bulkFinalizeFill(Kokkos::View<AtomicPairCounter*,KokkosExecSpace> const apc,const int& threadId) {
+    auto m = apc(0).get().m;
+    auto n = apc(0).get().n;
+    if (m >= nbins()) {  // overflow!
+      off[nbins()] = uint32_t(off[nbins() - 1]);
+      return;
+    }
+    auto i = m + threadId;
+    if(i < totbins()){
+      off[i] = n;
+    }
+  }
+
+  KOKKOS_INLINE_FUNCTION void count(T t) {
+    uint32_t b = bin(t);
+    assert(b < nbins());
+    atomicIncrement(off[b]);
+  }
+
+  KOKKOS_INLINE_FUNCTION void fill(T t, index_type j) {
+    uint32_t b = bin(t);
+    assert(b < nbins());
+    auto w = atomicDecrement(off[b]);
+    assert(w > 0);
+    bins[w - 1] = j;
+  }
+
+  KOKKOS_INLINE_FUNCTION void count(T t, uint32_t nh) {
+    uint32_t b = bin(t);
+    assert(b < nbins());
+    b += histOff(nh);
+    assert(b < totbins());
+    atomicIncrement(off[b]);
+  }
+
+  KOKKOS_INLINE_FUNCTION void fill(T t, index_type j, uint32_t nh) {
+    uint32_t b = bin(t);
+    assert(b < nbins());
+    b += histOff(nh);
+    assert(b < totbins());
+    auto w = atomicDecrement(off[b]);
+    assert(w > 0);
+    bins[w - 1] = j;
+  }
+
+  #pragma hd_warning_disable
+  template<typename Histo>
+  static KOKKOS_INLINE_FUNCTION void finalize(Kokkos::View<Histo*,KokkosExecSpace> histo) {
+    // assert(off[totbins() - 1] == 0);
+    // for(uint32_t i=0;i<totbins();++i)
+    //   printf("1 %04i off[%04i] = %04i\n",teamMember.team_rank(),i,off[i]);
+    Kokkos::parallel_scan("finalize",Histo::totbins(),
+      KOKKOS_LAMBDA(const int& i, uint32_t& update, const bool& final){
```

Comment:
```suggestion
      KOKKOS_LAMBDA(const int i, uint32_t& update, const bool final){
```

Reply:
The Kokkos Documentation for `parallel_scan` lambdas uses pass by reference for all three variables in the function. Do you have a reason we should not follow that convention? I figured it would avoid a copy.

Reply:
Just that usually taking primitive types by value is faster than going through the pointer (which reference internally is). E.g. C++ core guidelines recommend that https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-in.